File: rules/gitlab/java/endpoint/test-WeakHostNameVerification.java
// License: LGPL-3.0 License (c) find-sec-bugs
package endpoint;

import javax.net.ssl.*;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;

public class WeakHostNameVerification {

    public void useAllHosts() {
        HttpsURLConnection.setDefaultHostnameVerifier(new AllHosts());
    }

    public void useTrustAllManager() throws NoSuchAlgorithmException, KeyManagementException {
        final TrustManager[] trustAllCerts = new TrustManager[] { new TrustAllManager() };
        final SSLContext sslContext = SSLContext.getInstance("SSL");
        sslContext.init(null, trustAllCerts, null);
        HttpsURLConnection.setDefaultSSLSocketFactory(sslContext.getSocketFactory());
    }

    public void useSecurityBypasser() {
        SecurityBypasser.destroyAllSSLSecurityForTheEntireVMForever();
    }
}

class AllHosts implements HostnameVerifier {
    public boolean verify(final String hostname, final SSLSession session) {
        return true;
    }
}

class TrustAllManager implements X509TrustManager {
    public void checkClientTrusted(final X509Certificate[] cert, final String authType)
            throws CertificateException {
    }

    public void checkServerTrusted(final X509Certificate[] cert, final String authType)
            throws CertificateException {
    }

    public X509Certificate[] getAcceptedIssuers() {
        return null;
    }
}

class SecurityBypasser {
    public static void destroyAllSSLSecurityForTheEntireVMForever() {
        try {
            final TrustManager[] trustAllCerts = new TrustManager[] { new TrustAllManager() };
            final SSLContext sslContext = SSLContext.getInstance("SSL");
            sslContext.init(null, trustAllCerts, null);
            HttpsURLConnection.setDefaultSSLSocketFactory(sslContext.getSocketFactory());
            HttpsURLConnection.setDefaultHostnameVerifier(new AllHosts());
        } catch (final NoSuchAlgorithmException e) {
            e.printStackTrace();
        } catch (final KeyManagementException e) {
            e.printStackTrace();
        }
    }

}

File: rules/gitlab/scala/endpoint/test-WeakHostNameVerification.scala
// License: LGPL-3.0 License (c) find-sec-bugs
package endpoint

import javax.net.ssl._
import java.security.KeyManagementException
import java.security.NoSuchAlgorithmException
import java.security.cert.CertificateException
import java.security.cert.X509Certificate


class WeakHostNameVerification {
  def useAllHosts(): Unit = {
    HttpsURLConnection.setDefaultHostnameVerifier(new AllHosts)
  }

  @throws[NoSuchAlgorithmException]
  @throws[KeyManagementException]
  def useTrustAllManager(): Unit = {
    val trustAllCerts = Array[TrustManager](new TrustAllManager)
    val sslContext = SSLContext.getInstance("SSL")
    sslContext.init(null, trustAllCerts, null)
    HttpsURLConnection.setDefaultSSLSocketFactory(sslContext.getSocketFactory)
  }

  def useSecurityBypasser(): Unit = {
    SecurityBypasser.destroyAllSSLSecurityForTheEntireVMForever()
  }
}

class AllHosts extends HostnameVerifier {
  def verify(hostname: String, session: SSLSession) = true
}

class TrustAllManager extends X509TrustManager {
  @throws[CertificateException]
  def checkClientTrusted(cert: Array[X509Certificate], authType: String): Unit = {
  }

  @throws[CertificateException]
  def checkServerTrusted(cert: Array[X509Certificate], authType: String): Unit = {
  }

  def getAcceptedIssuers(): Array[X509Certificate] = null
}

class TrustAllManagerTwo extends X509TrustManager {
  @throws[CertificateException]
  def checkClientTrusted(cert: Array[X509Certificate], authType: String): Unit = {
  }

  @throws[CertificateException]
  def checkServerTrusted(cert: Array[X509Certificate], authType: String): Unit = {
  }

  def getAcceptedIssuers(): Array[X509Certificate] = {
    Array()
  }
}

object SecurityBypasser {
  def destroyAllSSLSecurityForTheEntireVMForever(): Unit = {
    try {
      val trustAllCerts = Array[TrustManager](new TrustAllManager)
      val sslContext = SSLContext.getInstance("SSL")
      sslContext.init(null, trustAllCerts, null)
      HttpsURLConnection.setDefaultSSLSocketFactory(sslContext.getSocketFactory)
      HttpsURLConnection.setDefaultHostnameVerifier(new AllHosts)
    } catch {
      case e: NoSuchAlgorithmException =>
        e.printStackTrace
      case e: KeyManagementException =>
        e.printStackTrace
    }
  }
}

